ldap principal provider: add option to add user to principals list

tierpod opened the pull request:
Hello. I missed some feature in signmykey, so I made this small patch. I think it can be useful for anybody else.

We have some servers integrated with ldap (Active Directory) via sssd. So it will be easier to integrate signmykey with such configuration if principals list contains domain username. In this case, we don't have to add AuthorizedPrincipalsFile to sshd_config.

pablo-ruth commented:
Hi @tierpod, that's a good idea! Do you need to have principals from ldap groups at the same time or another "Principals" provider adding only the username can be sufficient ?

pablo-ruth commented:
@c35sys @HammerZ3it what do you think about adding the username by default to the principals list? Since it's the default mode of SSH CA to check for the username if there is no AuthorizedPrincipalsFile defined, it seems fine to add it by default?

tierpod commented:
> Do you need to have principals from ldap groups at the same time

Actually, I thought about it. In case of sssd+ldap integration, ldap username is enough. But I also tested the case when I provided access based on principal from domain group (on external server without sssd). But I didn't decide yet, we are currently in testing process.

> another "Principals" provider adding only the username can be sufficient

It's interesting idea. I can take a look if you want.

And maybe it will be useful to use many principal providers at the same time? So administrators will be able to adjust the configuration for their needs

tierpod commented:
I played around this idea:

- added simple "user" principals provider as you described above
- added ability to configure multiple principals providers (struggled some time with viper, it's much harder than yaml unmarshalling), configuration section may look like this:
  ```yaml
  principalProviders:
    user:
    ldap:
      ldapAddr: ...
  ```
- Improved logging in principals providers

I'm going to improve my code in the next few days and make new PR.

c35sys commented:
> @c35sys @HammerZ3it what do you think about adding the username by default to the principals list? Since it's the default mode of SSH CA to check for the username if there is no AuthorizedPrincipalsFile defined, it seems fine to add it by default?

It can be a nice to have indeed, and it should be an option to not change current deployments, even if it is the enable by default.

pablo-ruth commented:
Ok thanks @c35sys. @tierpod can you just set the "addUser" config to true by default in your PR? Otherwise LGTM.

tierpod commented:
@pablo-ruth I made another PR https://github.com/signmykeyio/signmykey/pull/84 with changes I described above. It's quite massive. You can look at it, and choose which do you like more

pablo-ruth commented:
I close this one in favor of #84 .
